### PR TITLE
Improve Provider generics

### PR DESCRIPTION
**Pull request by Robo210**
The generics around the Provider struct were confusing. This PR will hopefully make things a lot clearer.

Old:
"Normal" output and Common Schema output were completely different implementations, even though they contained a lot of duplicate code. This meant that everything that touched a provider needed to be parameterized around the provider type and which implementation they were being instantiated for. This was done via associated types on the ProviderTypes trait, which further necessitated extra constraints on the associated types everywhere the ProviderTypes trait was used. This all gets very verbose and hard to follow.

New:
There is now one implementation of most provider methods. The output trait (EventWriter) is now parameterized over the output format (normal, common schema), and the provider has multiple implementations of the trait based on constraints on the output format generic. This reduces code duplication and final binary size, and should be easier to follow when reading the code.

Other changes:
- Store the provider name as a Box<str> rather than a String. This is unnecessary, but it keeps the name immutable.
- Remove the use of provider group name as a target filter on linux. I can't think of any way that this would actually be useful, and it was causing a split in behavior between Windows and Linux.
- Removed the common_schema feature and just made it always available. If you don't use it, the linker should be able to remove it for you.

### CODE REVIEW

File: src/layer/filter.rs [hunk 1]
```diff
@@ -2,17 +2,15 @@ use tracing::Subscriber;
...
 
-use super::*;
+use super::EtwFilter;
 
-impl<S, Mode> Filter<S> for EtwFilter<S, Mode>
+impl<S, OutMode: OutputMode> Filter<S> for EtwFilter<S, OutMode>
 where
     S: Subscriber + for<'a> LookupSpan<'a>,
-    Mode: ProviderTypes + 'static,
-    Mode::Provider: EventWriter<Mode> + 'static,
```
Comment (Robo210) [inline src/layer/filter.rs:15]: These 2 constraints in the old code were the main source of confusion and difficulty following the internal logic.

Comment (Robo210) [reply src/layer/filter.rs:15]: `Mode: ProviderTypes` - the Mode generic type implements the ProviderTypes trait. `Mode::Provider: EventWriter<Mode>` - the `Provider` associated type of the ProviderTypes trait implements EventWriter<Mode> trait. EventWriter will then have the same two constraints on its generics.

File: src/layer/filter.rs [hunk 2]
```diff
@@ -25,7 +23,7 @@ where
             self.layer.default_keyword
         };
 
-        if Mode::supports_enable_callback() {
+        if crate::native::Provider::<OutMode>::supports_enable_callback() {
```
Comment (Robo210) [inline src/layer/filter.rs:26]: Using the exported type rather than trying to put everything into generics helped reduce the complexity a lot. The drawback is it makes hacking up a pseudo cross-compilation environment in VSCode harder.

File: src/layer/mod.rs [hunk 3]
```diff
@@ -10,23 +10,17 @@ use tracing_core::callsite;
...
 
 use crate::{
-    native::{EventWriter, ProviderTypes},
+    native::{OutputMode, ProviderTraits},
     statics::get_event_metadata,
 };
 
-pub(crate) struct _EtwLayer<S, Mode: ProviderTypes>
-where
-    Mode::Provider: crate::native::EventWriter<Mode> + 'static,
```
Comment (Robo210) [inline src/layer/mod.rs:19]: These constraints were necessary only because the compiler can't look through nested classes to determine if constraints on the bottommost class are satisfied by the generics on the topmost instantiation. Each class must replicate the constraints all the way up the chain. Removing all of this makes the code easier to follow.

File: src/native/etw.rs [hunk 4]
```diff
@@ -342,3 +351,264 @@ impl super::EventWriter<Provider> for Provider {
         });
     }
 }
+
+struct CommonSchemaPartCBuilder<'a> {
```
Comment (Robo210) [inline src/native/etw.rs:355]: This can probably be moved to a common location to be shared between etw and user_events.

**Comment by Robo210** [comment]
On the test binary with both normal and cs output layers, the final binary is 6 kilobytes smaller on disk.

**Review by idigdoug — APPROVED** [review]
